### PR TITLE
expression: use precision and frac from MyDecimal instead of datum for encoding

### PR DESCRIPTION
**Pull request by gengliqi**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/42622

Problem Summary:

### What is changed and how it works?
The bug was caused by a error when a decimal datum is encoded because the precision and fraction from datum are different from the ones in datum.MyDecimal. The precision and fraction from MyDecimal should be used for encoding.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: expression/expr_to_pb.go [hunk 1]
```diff
@@ -129,7 +129,8 @@ func (pc *PbConverter) encodeDatum(ft *types.FieldType, d types.Datum) (tipb.Exp
 	case types.KindMysqlDecimal:
 		tp = tipb.ExprType_MysqlDecimal
 		var err error
-		val, err = codec.EncodeDecimal(nil, d.GetMysqlDecimal(), d.Length(), d.Frac())
+		// use precision and frac from MyDecimal instead of datum
```
Comment (windtalker) [inline expression/expr_to_pb.go:132]: Add more comments about this?

File: util/chunk/row.go [hunk 2]
```diff
@@ -173,14 +173,15 @@ func (r Row) DatumWithBuffer(colIdx int, tp *types.FieldType, d *types.Datum) {
...
+			dec := r.GetMyDecimal(colIdx)
+			d.SetMysqlDecimal(dec)
 			d.SetLength(tp.GetFlen())
 			// If tp.decimal is unspecified(-1), we should set it to the real
 			// fraction length of the decimal value, if not, the d.Frac will
 			// be set to MAX_UINT16 which will cause unexpected BadNumber error
 			// when encoding.
 			if tp.GetDecimal() == types.UnspecifiedLength {
-				d.SetFrac(d.Frac())
+				d.SetFrac(int(dec.GetDigitsFrac()))
```
Comment (windtalker) [inline util/chunk/row.go:184]: Any test case that will trigger this bug?

File: expression/expr_to_pb.go (hunk 1 above)
Comment (gengliqi) [reply expression/expr_to_pb.go]: Added.

File: util/chunk/row.go (hunk 2 above)
Comment (gengliqi) [reply util/chunk/row.go:184]: It seems difficult because almost everywhere the `MyDecimal` parameter is set, the `Frac` parameter is set.

**Review by windtalker — APPROVED** [review]
LGTM

File: expression/integration_test/integration_test.go [hunk 3]
```diff
@@ -2059,3 +2059,25 @@ func TestIssue41986(t *testing.T) {
...
+	where exists (
+		select 1 as c0
+		union all
+		select
+			1 as c0
+		  from
+			(t0 as ref_88)
+		  where (subq_0.c4) >= (subq_0.c4)
+		 );`).Check(testkit.Rows("23.0000"))
+}
```
Comment (SeaRise) [inline expression/integration_test/integration_test.go:2083]: please moving to https://github.com/pingcap/tidb/blob/df7631965cb6197eb960f828118f69fbb4f1788a/tests/integrationtest/t/expression/issues.test

**Review by SeaRise — COMMENTED** [review]
Others LGTM

File: tests/integrationtest/t/expression/issues.test [hunk 4]
```diff
@@ -1888,3 +1888,11 @@ select (@j := case when substr(t2.b,1,3)=@i then 1 else @j+1 end) from t2, (sele
...
 select cast(b'10000000' as unsigned) / 10;
 select b'10000000' / 10;
+
+# TestIssue42622
+drop table if exists t0;
+create table t0 (c0 int);
+select * from (select (92 / 4) as c4) as subq_0 where exists (
+    select 1 as c0
+		union all
+		select 1 as c0 from (t0 as ref_88) where (subq_0.c4) >= (subq_0.c4));
```
Comment (SeaRise) [inline tests/integrationtest/t/expression/issues.test]:
```suggestion
    select 1 as c0
    union all
    select 1 as c0 from (t0 as ref_88) where (subq_0.c4) >= (subq_0.c4));
```

File: tests/integrationtest/t/expression/issues.test [hunk 5]
```diff
@@ -1888,3 +1888,11 @@ select (@j := case when substr(t2.b,1,3)=@i then 1 else @j+1 end) from t2, (sele
...
 select cast(b'10000000' as unsigned) / 10;
 select b'10000000' / 10;
+
+# TestIssue42622
+drop table if exists t0;
+create table t0 (c0 int);
+select * from (select (92 / 4) as c4) as subq_0 where exists (
+  select 1 as c0
+  union all
+	select 1 as c0 from (t0 as ref_88) where (subq_0.c4) >= (subq_0.c4));
```
Comment (SeaRise) [inline tests/integrationtest/t/expression/issues.test:1898]:
```suggestion
select 1 as c0
union all
select 1 as c0 from (t0 as ref_88) where (subq_0.c4) >= (subq_0.c4));
```

**Review by SeaRise — APPROVED** [review]
LGTM

**Pull request merged**